File: 0260-single-number-iii/0260-single-number-iii.php
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer[]
     */
    function singleNumber($nums) {
        $x = 0;
        for($i = 0; $i < sizeof($nums); $i++)
            $x ^= $nums[$i];
        $aX = $x;
        $p = 0;
        while(true)
        {
            if($aX % 2 != 0)
                break;
            $p += 1;
            $aX /= 2;
        }
        $n1 = 0;
        $n2 = 0;
        for($i = 0; $i < sizeof($nums); $i++)
        {
            if(($nums[$i] & 2 ** $p) == 0)
                $n1 ^= $nums[$i];
            else
                $n2 ^= $nums[$i];
        }
        return [$n1, $n2];
    }
}
    

File: 0260-single-number-iii/0260-single-number-iii.rb
# @param {Integer[]} nums
# @return {Integer[]}
def single_number(nums)
    x = 0
    nums.each do |num|
        x ^= num
    end
    aX = x
    p = 0
    while true
        if aX % 2 != 0
            break
        end
        p += 1
        aX /= 2
    end
    n1, n2 = 0, 0
    nums.each do |num|
        if num & 2 ** p == 0
            n1 ^= num
        else
            n2 ^= num
        end
    end
    return Array.new([n1, n2])
end